### PR TITLE
democluster: tracing fixes

### PR DESCRIPTION
**Pull request by andreimatei**
See individual commits.

### CODE REVIEW

File: pkg/cli/democluster/demo_cluster.go [hunk 1]
```diff
@@ -584,7 +586,12 @@ func (c *transientCluster) startNodeAsync(
 	serv := c.servers[idx]
 	tag := fmt.Sprintf("start-n%d", idx+1)
 	return c.stopper.RunAsyncTask(ctx, tag, func(ctx context.Context) {
-		ctx = logtags.AddTag(ctx, tag, nil)
+		// We call Start() with context.Background() because we don't want the
+		// tracing span corresponding to the task started just above to leak into
+		// the new server. Server's have their own Tracers, different from the one
+		// used by this transientCluster, and so traces inside the Server can't be
+		// combined with traces from outside it.
+		ctx = logtags.AddTag(context.Background(), tag, nil)
```
Comment (knz) [inline pkg/cli/democluster/demo_cluster.go:594]: at the very least, do
```go
ctx = logtags.WithTags(context.Background(), logtags.FromContext(ctx))
ctx = logtags.AddTag(ctx, tag, nil)
```

but don' you want an `AnnotateCtx()` as well, to connect it to the tracer?

File: pkg/cli/democluster/demo_cluster.go [hunk 2]
```diff
@@ -386,11 +386,13 @@ func (c *transientCluster) Start(
 			for i := 0; i < c.demoCtx.NumNodes; i++ {
 				latencyMap := c.servers[i].Cfg.TestingKnobs.Server.(*server.TestingKnobs).ContextTestingKnobs.ArtificialLatencyMap
 				c.infoLog(ctx, "starting tenant node %d", i)
+				stopper := stop.NewStopper()
+				c.stopper.AddCloser(stop.CloserFn(func() { stopper.Stop(context.Background()) }))
```
Comment (knz) [inline pkg/cli/democluster/demo_cluster.go:390]: can you run this `AddCloser` after `StartTenant` has completed, then do:

```go
c.stopper.AddCloser(... func {
   ctx := c.servers[i].AmbientCtx.AnnotateCtx(context.Background())
   stopper.Stop(ctx)
})
```

**Review by andreimatei — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/72924)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @knz)

___
*[pkg/cli/democluster/demo_cluster.go, line 390 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/72924#-MonzyAb63FLmTW-_xuy-r2-390:-MooDUBC5CuhqGrCUTbL:b-tf5wgv) ([raw file](https://github.com/cockroachdb/cockroach/blob/dd2b9a545c8baf431fbc629fa964614dd317a8ab/pkg/cli/democluster/demo_cluster.go#L390)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

can you run this `AddCloser` after `StartTenant` has completed, then do:

```go
c.stopper.AddCloser(... func {
   ctx := c.servers[i].AmbientCtx.AnnotateCtx(context.Background())
   stopper.Stop(ctx)
})
```
</blockquote></details>

I don't think you want `c.servers[i]`, I think you want the tenant's  `AmbientCtx`, right? Cause stopping the tenant doesn't have much to do with stopping the server, I think.
See now.
___
*[pkg/cli/democluster/demo_cluster.go, line 594 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/72924#-MonzyAb63FLmTW-_xuy-r2-594:-MooE9gb1uZoQWGZESY0:bh82pbw) ([raw file](https://github.com/cockroachdb/cockroach/blob/dd2b9a545c8baf431fbc629fa964614dd317a8ab/pkg/cli/democluster/demo_cluster.go#L594)):*
> at the very least, do

done

> but don' you want an AnnotateCtx() as well, to connect it to the tracer?

I'm not sure I follow this. Which `AmbientCtx` are you suggesting to use - `serv`'s? I don't know about that; `[Test]Server.Start()` seems like an entry point to me; it should do whatever annotations it wants to do itself.
Also, `AnnotateCtx()` doesn't deal with spans, so I'm not sure about "connect it to the tracer" (and just in case this wasn't here, the whole point here is to _disconnect_ this `Server.Start()` from the `c.stopper`'s `Tracer`). There's an `AnnotateCtxWithSpan()`, but similarly I don't feel a need for that.

**Review by knz — APPROVED** [review]

File: pkg/cli/democluster/demo_cluster.go [hunk 3]
```diff
@@ -584,7 +592,14 @@ func (c *transientCluster) startNodeAsync(
 	serv := c.servers[idx]
 	tag := fmt.Sprintf("start-n%d", idx+1)
 	return c.stopper.RunAsyncTask(ctx, tag, func(ctx context.Context) {
+		// We call Start() with context.Background() because we don't want the
+		// tracing span corresponding to the task started just above to leak into
+		// the new server. Server's have their own Tracers, different from the one
+		// used by this transientCluster, and so traces inside the Server can't be
+		// combined with traces from outside it.
+		ctx = logtags.WithTags(context.Background(), logtags.FromContext(ctx))
```
Comment (knz) [inline pkg/cli/democluster/demo_cluster.go:600]: this means the code is assuming that the Start() call will have its own tracer, right?

Comment (knz) [reply pkg/cli/democluster/demo_cluster.go:600]: if so, maybe worth clarifying in the comment

**Review by andreimatei — COMMENTED** [review]
please look again at the first commit

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/72924#-:-MotBNkQAjg9zj-ViKYp:by41heb)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @knz)

___
*[pkg/cli/democluster/demo_cluster.go, line 600 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/72924#-MonzyAb63FLmTW-_xuy-r4-600:-MotB4rR2cd11qKzSp6y:b913dje) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d0db12a83710dcf259d574ccbcf98125e0f3925/pkg/cli/democluster/demo_cluster.go#L600)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

if so, maybe worth clarifying in the comment
</blockquote></details>

well, this code here is not particularly "assuming" anything. It doesn't care if there's tracing or if there isn't. I've thought about saying something, but I'm not sure exactly what to say beyond the note that we already have here that hints at the server's tracer.

**Comment by andreimatei** [comment]
Rafa please give this guy a stamp when you can

**Review by knz — APPROVED** [review]
> Reviewed 1 of 1 files at r5, 1 of 1 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/72924)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei)

**Comment by andreimatei** [comment]
TFTR

bors r+

**Pull request merged**